packages/perseus-editor/src/widgets/free-response-editor.tsx

```diff
@@ -32,18 +39,148 @@ class FreeResponseEditor extends React.Component<Props> {
         return warnings;
     }
 
+    handleUpdateCriterion = (
+        index: number,
+        criterion: PerseusFreeResponseWidgetScoringCriterion,
+    ) => {
+        const newCriteria = this.props.scoringCriteria.map((c, i) => {
+            if (i === index) {
+                return criterion;
+            }
+            return c;
+        });
+
+        this.props.onChange({
+            scoringCriteria: newCriteria,
+        });
+    };
+
+    handleDeleteCriterion = (index: number) => {
+        this.props.onChange({
+            scoringCriteria: this.props.scoringCriteria.filter(
+                (_, i) => i !== index,
+            ),
+        });
+    };
+
+    handleAddCriterion = () => {
+        this.props.onChange({
+            scoringCriteria: [...this.props.scoringCriteria, {text: ""}],
+        });
+    };
+
+    renderCriteriaList = () => {
+        return this.props.scoringCriteria.map((criterion, index) => {
+            return (
+                <CriterionEditor
+                    criterion={criterion}
+                    index={index}
+                    key={index}
```

Comment:
I get nervous when I see [key set to index](https://react.dev/learn/rendering-lists#why-does-react-need-keys), but I imagine this might be unavoidable because in the editor maybe we can't use `criterion.text`? Because (1) it might be an empty string and (2) we might have multiple empty strings until we fill out the form?

Reply:
That's right. We can do `${index}-${criterion.text}`, though.